### DISCUSSION TITLE
Huawei P40 Pro Disconnects from 5G SA Network After a Couple Seconds of Internet Connection

### DISCUSSION THREAD

**Discussion by Lusani1** [srsran/srsRAN_Project#73] [General Help] [answered]
Dear All,

May you kindly assist with diagnosing the root cause of our 5G testbed instability.

**Our setup is as follows:**

- 5G SA Indoor testbed using srsRAN Project
- PLMN 20895
- Band 3 (Note: we do not yet have an RF license.)
- B210 and X300 SDRs (we tested with both)
- Open5GS Core Network (NSSAI>>SST=1,SD=0)
- 1GB NICs on the backhaul interface
- PCIe on fronthaul for tests using X300 and USB 3.0 for B210
- RAN server is an i7 with 12 CPUs and 32GB RAM

Our issue: When connecting the UE (Huawei P40 Pro), the network is able to establish a PDU session and an Internet connection. Unfortunately, the phone is disconnected after a couple of minutes (sometimes seconds) of Internet browsing.

**We have tried debugging the issue as follows:**

- Set the RAN server to performance mode and disables c-states
- Tweaked the Tx and Rx gains
- Killed all processes not needed by the RAN
- Attenuated the Rx and Tx to 0
- Tried testing on other NR bands 
- Tried testing using OAI gNB but still faced the same issue.

**Our suspicion:**
The MCS and CQI values are not consistent likely due to poor radio conditions. 

The logs are available here >>  [gnb.log](https://github.com/srsran/srsRAN_Project/files/11420223/gnb.log).
The NGAP and MAC layer pcaps are available at [this](https://drive.google.com/drive/folders/1I_jNYROF-HugcdjCmlDseMoYo_Ihe1rA?usp=share_link) link.
The example UE console traces are attached below:
![UE_trace_works_but_poor_cqi](https://user-images.githubusercontent.com/116084529/236800877-6d371ff7-c8e5-4747-a07d-9d5e55fdc45e.png)
![UE_trace_works_but_poorer_cqi](https://user-images.githubusercontent.com/116084529/236800903-b3a0d636-26a9-4e18-b0ed-767f0ae87945.png)

**Comment by andrepuschmann** [comment]
Hey, your RF conditions aren't ideal. You have to tweak your Rx/Tx gains. Note that the B200 and X300 need very different gain settings. I suggest you stick with one device, optimize that and then use the other. Please also follow our COTS UE appnote. When you send logs again, please set the `all_level` to info.

**Reply by HeroesLament** [reply]
It may be more evident to people not read into LTE-speak that PUSCH relates to the receive power of the UE. Might be worth to add a dBm unit to the column header?

**Reply by ismagom** [reply]
That's actually SNR (in dB).

**Comment by ismagom** [comment]
Hi @Lusani1 , can you send logs in info mode please? `all_level = info` in the config file, log section. Thanks!

**Comment by ralfkundel** [comment]
Hi,
in your pcap (I cannot parse it but I checked the hexstring) I see two PDU requests: ims and internet. If the Huawei P40 does not get an ims PDU session it will disconnect after a few seconds (at least ours).
Details: https://www.kom.tu-darmstadt.de/papers/KMK+22.pdf

**Comment by Lusani1** [comment]
Thank you so much for the diagnosis and the paper. I will definitely go
through it.

To solve the problem, do I need to configure an additional APN in the CN?

Regards,
Lusani

**Comment by ralfkundel** [comment]
Hi,
I remember two workarounds:
1. in free5gc you could set IMS disabled/enabled by setting imsVoPS to 0/1 https://github.com/free5gc/free5gc/blob/main/config/amfcfg.yaml
if you disable it, it takes ~3 minutes until the P40 connected because after the second or third attempt it accepted an 5G radio without VoIP capabilities.
2. you create an "ims" PDU including data network. you must configure in your core somewhere the ims server IP the huawei is connecting to.  at this IP you would need an ims server or at least an ims fake server (just open a tcp port in python and respond some "garbage data" - the huawei will fail the ims handshake but stay connected in internet)
3. use another phone ;-) e.g. asus zenfone or One+ I had good experience with (but not yet tested with srsRAN)

best,
Ralf

**Reply by robertfalkenberg** [reply, marked as answer]
@Lusani1 this also happens with Google Pixel 6 and is related to missing IMS. A quick workaround without the need to setup IMS is here: https://github.com/srsran/srsRAN_Project/discussions/100#discussioncomment-6260974
